prusti-viper/src/encoder/backtranslation.rs

```diff
@@ -0,0 +1,457 @@
+use std::collections::HashMap;
+use std::convert::TryFrom;
+
+use viper::silicon_counterexample::*;
+use viper::VerificationError;
+use prusti_interface::data::ProcedureDefId;
+use prusti_common::vir::CfgMethod;
+use crate::encoder::Encoder;
+use crate::encoder::places::{Local, LocalVariableManager, Place};
+use crate::encoder::counterexample::*;
+
+use rustc_middle::mir;
+use rustc_middle::ty::{self, Ty, AdtKind, AdtDef, TyCtxt};
+use rustc_span::MultiSpan;
+
+
+pub struct Backtranslator<'tcx>{
+    mir: mir::Body<'tcx>,
+    def_id: ProcedureDefId,
+    silicon_counterexample: SiliconCounterexample,
+    tcx: TyCtxt<'tcx>,
+    is_pure: bool,
+    disc_info: HashMap<(ProcedureDefId, String), Vec<String>>,
+}
+
+impl<'tcx> Backtranslator<'tcx> {
+    pub fn new(
+        encoder: &Encoder<'_, 'tcx>,
+        def_id: ProcedureDefId,
+        silicon_opt_counterexample: Option<SiliconCounterexample>
+    ) -> Option<Backtranslator<'tcx>> {
+        if let Some(silicon_counterexample) = silicon_opt_counterexample {
+            let mir = encoder.env().get_procedure(def_id).get_mir().clone();
+            let tcx = encoder.env().tcx();
+            let is_pure = encoder.is_pure(def_id);
+            let disc_info = encoder.discriminants_info();
+            Some(Backtranslator{
+                mir,
+                def_id,
+                silicon_counterexample,
+                tcx,
+                is_pure,
+                disc_info,
+            })
+        } else {
+            None
+        }
+    }
+}
+
+
+pub fn backtranslate<'tcx>(
+    encoder: &Encoder,
+    def_id: ProcedureDefId,
+    silicon_counterexample: Option<SiliconCounterexample>,
+) -> Counterexample {
+    let opt_backtranslator = Backtranslator::new(encoder, def_id, silicon_counterexample);
+    if let Some(backtranslator) = opt_backtranslator {
+        //get all the needed information from the mir
+        let var_debug_info = backtranslator.mir.var_debug_info.clone();
+        let local_variable_manager = LocalVariableManager::new(&backtranslator.mir.local_decls);
+        let arg_count = backtranslator.mir.arg_count;
+
+        //optimally (at a later stage) we would use the "main" counterexample 
+        //from silicon, the one not associated with any label, because it contains
+        //the values of the function when it fails. But currently
+        //most values can not be obtained there because they're folded
+        let last_label: Option<&String> = backtranslator
+            .silicon_counterexample
+            .label_order.last();
+        //optimally this label would just be "old", but as of now mostly the values
+        //are not available at this point
+        let old_label = String::from("l0");
+
+
+
```

Comment:
Let's not have these big gaps. If it makes sense you can split the function into smaller ones.